File: src/typescript/src/xiaoyi/memory/stm/cache.ts
/**
 * # STM Cache
 *
 * `cache` provides caching layer for STM with LRU eviction.
 *
 * Path: `xiaoyi.memory.stm.cache`
 *
 * @module memory.stm.cache
 * @brief LRU cache for STM entries
 * @group Memory
 * @since 0.1.0
 * @author Miruamel
 * @see memory.stm
 * @see memory.stm.context
 * @security Cache may contain sensitive conversation data. Clear on logout.
 */
import { StmEntry } from "..";

/**
 * Cache entry with access tracking.
 *
 * @brief Cached STM entry with LRU metadata
 * @group Memory
 * @since 0.1.0
 */
export interface CacheEntry<T = StmEntry> {
  /** Cached value. */
  value: T;
  /** Last access timestamp. */
  lastAccess: number;
  /** Access count. */
  accessCount: number;
}

/**
 * LRU cache implementation.
 *
 * @brief Least-recently-used cache for STM
 * @group Memory
 * @since 0.1.0
 * @threadsafe
 * @example
 * ```typescript
 * const cache = new LruCache<StmEntry>({ maxSize: 1000 });
 * cache.set("key", entry);
 * const entry = cache.get("key");
 * ```
 */
export class LruCache<T = StmEntry> {
  private cache = new Map<string, CacheEntry<T>>();
  private readonly _maxSize: number;
  public readonly maxSize: number;
  private accessOrder: string[] = []; // most recent first

  /**
   * Create LRU cache.
   *
   * @param options - Cache options
   * @since 0.1.0
   */
  constructor(options: { maxSize?: number } = {}) {
    this._maxSize = options.maxSize ?? 1000;
    this.maxSize = this._maxSize;
  }

  /**
   * Get value by key.
   *
   * @param key - Cache key
   * @returns Value or undefined
   * @since 0.1.0
   */
  get(key: string): T | undefined {
    const entry = this.cache.get(key);
    if (!entry) return undefined;

    // Move to front (most recent)
    this.moveToFront(key);
    entry.lastAccess = Date.now();
    entry.accessCount++;
    return entry.value;
  }

  /**
   * Set value.
   *
   * @param key - Cache key
   * @param value - Value to cache
   * @since 0.1.0
   */
  set(key: string, value: T): void {
    if (this._maxSize === 0) return; // No storage allowed

    const exists = this.cache.has(key);

    if (!exists && this.cache.size >= this._maxSize) {
      this.evict();
    }

    this.cache.set(key, {
      value,
      lastAccess: Date.now(),
      accessCount: exists ? this.cache.get(key)!.accessCount + 1 : 1,
    });

    // Move to front (most recent)
    this.moveToFront(key);
  }

  /**
   * Check if key exists.
   *
   * @param key - Cache key
   * @returns true if exists
   * @since 0.1.0
   */
  has(key: string): boolean {
    return this.cache.has(key);
  }

  /**
   * Delete key.
   *
   * @param key - Cache key
   * @returns true if deleted
   * @since 0.1.0
   */
  delete(key: string): boolean {
    const deleted = this.cache.delete(key);
    if (deleted) {
      this.removeFromOrder(key);
    }
    return deleted;
  }

  /**
   * Clear all entries.
   *
   * @since 0.1.0
   */
  clear(): void {
    this.cache.clear();
    this.accessOrder = [];
  }

  /**
   * Get cache size.
   *
   * @returns Number of entries
   * @since 0.1.0
   */
  get size(): number {
    return this.cache.size;
  }

  /**
   * Get all keys in LRU order (most recent first).
   *
   * @returns Array of keys
   * @since 0.1.0
   */
  keys(): string[] {
    return [...this.accessOrder];
  }

  /**
   * Get all entries in LRU order (most recent first).
   *
   * @returns Array of [key, value] pairs
   * @since 0.1.0
   */
  entries(): [string, T][] {
    return this.accessOrder.map((k) => [k, this.cache.get(k)!.value]);
  }

  /**
   * Get all values in LRU order (most recent first).
   *
   * @returns Array of values
   * @since 0.1.0
   */
  values(): T[] {
    return this.accessOrder.map((k) => this.cache.get(k)!.value);
  }

  /**
   * Move key to front (most recent).
   */
  private moveToFront(key: string): void {
    this.removeFromOrder(key);
    this.accessOrder.unshift(key);
  }

  /**
   * Remove key from access order.
   */
  private removeFromOrder(key: string): void {
    const idx = this.accessOrder.indexOf(key);
    if (idx >= 0) {
      this.accessOrder.splice(idx, 1);
    }
  }

  /**
   * Evict least recently used entry.
   *
   * @since 0.1.0
   */
  private evict(): void {
    if (this.accessOrder.length > 0) {
      const lruKey = this.accessOrder.pop()!;
      this.cache.delete(lruKey);
    }
  }

  /**
   * Get cache statistics.
   *
   * @returns Cache stats
   * @since 0.1.0
   */
  stats(): { size: number; maxSize: number; hitRate: number } {
    let totalAccess = 0;
    for (const entry of this.cache.values()) {
      totalAccess += entry.accessCount;
    }

    return {
      size: this.cache.size,
      maxSize: this._maxSize,
      hitRate: totalAccess > 0 ? totalAccess / this.cache.size : 0,
    };
  }
}

/**
 * Create default STM cache.
 *
 * @returns Default LRU cache
 * @since 0.1.0
 * @group Memory
 */
export function createStmCache(): LruCache<StmEntry> {
  return new LruCache({ maxSize: 1000 });
}

File: src/rust/src/xiaoyi/memory/stm/cache.rs
//! # STM Cache
//!
//! `cache` provides a thread-safe LRU cache with TTL support for
//! short-term memory storage.
//!
//! Path: `xiaoyi::memory::stm::cache`
//!
//! - Layer 0: `memory`
//! - Layer 1: `stm`
//! - Layer 2: `cache` — LRU cache implementation.
//!
//! @module memory::stm::cache
//! @brief Thread-safe LRU cache with TTL
//! @group Memory
//! @since 0.1.0
//! @author Miruamel
//! @see crate::memory::stm
//! @see crate::memory::ltm::vector
//!
//! # Example
//!
//! ```rust
//! use xiaoyi::memory::stm::cache::LruCache;
//! use std::time::Duration;
//!
//! let cache = LruCache::new(100);
//! cache.insert("key".to_string(), "value".to_string(), Some(Duration::from_secs(60)));
//! let value = cache.get("key");
//! ```
//!
//! # Features
//!
//! - O(1) get/insert with LRU eviction.
//! - Optional TTL per entry.
//! - Thread-safe with RwLock.
//! - Configurable capacity.
//!
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::time::{Duration, Instant};
/// Cache entry with value and expiry.
///
/// @brief Stores cached value with optional TTL expiry
/// @group Memory
/// @since 0.1.0
#[derive(Debug, Clone)]
pub struct CacheEntry<V> {
    /// Cached value.
    pub value: V,
    /// Optional expiry timestamp.
    pub expiry: Option<Instant>,
}

/// Cache statistics.
///
/// @brief Cache performance metrics
/// @group Memory
/// @since 0.1.0
#[derive(Debug, Clone, Default)]
pub struct CacheStats {
    /// Number of cache hits.
    pub hits: usize,
    /// Number of cache misses.
    pub misses: usize,
    /// Number of LRU evictions.
    pub evictions: usize,
    /// Current cache size.
    pub size: usize,
    /// Maximum cache capacity.
    pub capacity: usize,
}

/// LRU cache with optional TTL.
///
/// @brief High-performance LRU cache with TTL support
/// @group Memory
/// @since 0.1.0
/// @see crate::memory::stm::buffer
#[derive(Debug)]
pub struct LruCache<K, V> {
    map: Arc<RwLock<HashMap<K, CacheEntry<V>>>>,
    order: Arc<RwLock<Vec<K>>>,
    capacity: usize,
    hits: Arc<RwLock<usize>>,
    misses: Arc<RwLock<usize>>,
    evictions: Arc<RwLock<usize>>,
}

impl<K, V> LruCache<K, V>
where
    K: Clone + Eq + std::hash::Hash + Send + Sync + 'static,
    V: Clone + Send + Sync + 'static,
{
    /// Create a new LRU cache with given capacity.
    ///
    /// @param capacity Maximum number of entries
    /// @return New LruCache instance
    /// @since 0.1.0
    pub fn new(capacity: usize) -> Self {
        Self {
            map: Arc::new(RwLock::new(HashMap::with_capacity(capacity))),
            order: Arc::new(RwLock::new(Vec::with_capacity(capacity))),
            capacity,
            hits: Arc::new(RwLock::new(0)),
            misses: Arc::new(RwLock::new(0)),
            evictions: Arc::new(RwLock::new(0)),
        }
    }

    /// Get a value by key.
    ///
    /// @param key Cache key
    /// @return Cloned value if present and not expired
    /// @since 0.1.0
    pub fn get(&self, key: &K) -> Option<V> {
        let mut map = self.map.write().unwrap();
        let mut order = self.order.write().unwrap();

        if let Some(entry) = map.get(key) {
            if let Some(expiry) = entry.expiry {
                if Instant::now() > expiry {
                    map.remove(key);
                    order.retain(|k| k != key);
                    *self.misses.write().unwrap() += 1;
                    return None;
                }
            }

            // Move to front (most recently used)
            order.retain(|k| k != key);
            order.insert(0, key.clone());

            *self.hits.write().unwrap() += 1;
            Some(entry.value.clone())
        } else {
            *self.misses.write().unwrap() += 1;
            None
        }
    }

    /// Insert a key-value pair with optional TTL.
    ///
    /// @param key Cache key
    /// @param value Value to store
    /// @param ttl Optional time-to-live
    /// @since 0.1.0
    pub fn insert(&self, key: K, value: V, ttl: Option<Duration>) {
        let mut map = self.map.write().unwrap();
        let mut order = self.order.write().unwrap();

        let expiry = ttl.map(|d| Instant::now() + d);

        if map.contains_key(&key) {
            order.retain(|k| k != &key);
        } else if map.len() >= self.capacity {
            // Evict LRU
            if let Some(lru_key) = order.pop() {
                map.remove(&lru_key);
                *self.evictions.write().unwrap() += 1;
            }
        }

        order.insert(0, key.clone());
        map.insert(key, CacheEntry { value, expiry });
    }

    /// Remove a key from cache.
    ///
    /// @param key Key to remove
    /// @return true if key was present
    /// @since 0.1.0
    pub fn remove(&self, key: &K) -> bool {
        let mut map = self.map.write().unwrap();
        let mut order = self.order.write().unwrap();

        let removed = map.remove(key).is_some();
        if removed {
            order.retain(|k| k != key);
        }
        removed
    }

    /// Clear all entries.
    ///
    /// @since 0.1.0
    pub fn clear(&self) {
        self.map.write().unwrap().clear();
        self.order.write().unwrap().clear();
    }

    /// Get current entry count.
    ///
    /// @return Number of entries
    /// @since 0.1.0
    pub fn len(&self) -> usize {
        self.map.read().unwrap().len()
    }

    /// Check if cache is empty.
    ///
    /// @return true if no entries
    /// @since 0.1.0
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Get cache capacity.
    ///
    /// @return Maximum capacity
    /// @since 0.1.0
    pub fn capacity(&self) -> usize {
        self.capacity
    }

    /// Clean up expired entries.
    ///
    /// @return Number of expired entries removed
    /// @since 0.1.0
    pub fn cleanup_expired(&self) -> usize {
        let mut map = self.map.write().unwrap();
        let mut order = self.order.write().unwrap();
        let now = Instant::now();

        let expired_keys: Vec<K> = map
            .iter()
            .filter(|(_, entry)| entry.expiry.map_or(false, |exp| now > exp))
            .map(|(k, _)| k.clone())
            .collect();

        for key in &expired_keys {
            map.remove(key);
            order.retain(|k| k != key);
        }

        expired_keys.len()
    }

    /// Get cache statistics.
    ///
    /// @return CacheStats with hits, misses, evictions, size, capacity
    /// @since 0.1.0
    pub fn stats(&self) -> CacheStats {
        CacheStats {
            hits: *self.hits.read().unwrap(),
            misses: *self.misses.read().unwrap(),
            evictions: *self.evictions.read().unwrap(),
            size: self.len(),
            capacity: self.capacity,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::thread::sleep;
    use std::time::Duration;

    #[test]
    fn test_basic_lru() {
        let cache = LruCache::new(3);
        cache.insert("a".to_string(), 1, None);
        cache.insert("b".to_string(), 2, None);
        cache.insert("c".to_string(), 3, None);

        assert_eq!(cache.get(&"a".to_string()), Some(1));
        assert_eq!(cache.get(&"b".to_string()), Some(2));
        assert_eq!(cache.get(&"c".to_string()), Some(3));

        // Access 'a' to make it MRU
        cache.get(&"a".to_string());

        // Insert 'd' should evict 'b' (LRU)
        cache.insert("d".to_string(), 4, None);
        assert_eq!(cache.get(&"b".to_string()), None);
        assert_eq!(cache.get(&"a".to_string()), Some(1));
        assert_eq!(cache.get(&"c".to_string()), Some(3));
        assert_eq!(cache.get(&"d".to_string()), Some(4));
    }
    #[test]
    fn test_ttl_expiry() {
        let cache = LruCache::new(10);
        cache.insert(
            "key".to_string(),
            "value".to_string(),
            Some(Duration::from_millis(50)),
        );
        assert_eq!(cache.get(&"key".to_string()), Some("value".to_string()));

        sleep(Duration::from_millis(100));
        assert_eq!(cache.get(&"key".to_string()), None);
    }
}
